README.md
# Sezzle Plugin for WooCommerce

## Introduction
This document will help you in installing `Sezzle` plugin in WooCommerce. This plugin is a certified one and listed [here](https://wordpress.org/plugins/sezzle-woocommerce-payment) in the marketplace.

## How to install the plugin?

* Login to `WordPress` admin.
* Make sure you have `WooCommerce` plugin installed.
* Go to `Plugins > Add New` and search for `Sezzle WooCommerce Payment` in the Search box.
* Once you find `Sezzle WooCommerce Payment` plugin, click on `install` and later `activate` the plugin.


## How to upgrade the plugin?

* Login to `WordPress` admin.
* Navigate to `Plugins > Installed Plugins`.
* You will see a notifcation below `Sezzle` plugin if any upgrade in available.
* Click on `Update now` to upgrade the plugin.

## Configure Sezzle

* To configure your `Sezzle` plugin in `WooCommerce`, complete the following steps. Prerequisite for this section is to obtain `Merchant ID`, `Private Key` and `Public Key` from [`Sezzle Merchant Dashboard`](https://dashboard.sezzle.in/merchant/). Sign Up if you have not signed up to get the necessities.
* Go to `WooCommerce > Settings > Payments > Sezzle`.
* Configure the plugin as follows:
    * Check the `Enable/Disable` checkbox for enabling `Sezzle`.
    * Check the `Payment option availability in other countries` if you want to allow `Sezzle` outside of `US` and `Canada`.
        * Just a note here, `Sezzle` operates only in `US` and `Canada`. Be sure to check this option.
    * Set `Merchant ID` as received from `Business` section of `Sezzle Merchant Dashboard`.
    * Set `Minimum Checkout Amount` if you want restrict `Sezzle` based on any particular amount.
    * Check the `Show Sezzle widget in product pages` checkbox for adding widget script in the Product Display Page which will help in enabling `Sezzle Widget` Modal in PDP.
    * Set `Private Key` as received from `API Keys` section of `Sezzle Merchant Dashboard`.
    * Set `Public Key` as received from your `API Keys` section of `Sezzle Merchant Dashboard`.
    * Use the `Sezzle API URL` as `https://gateway.sezzle.com/v1` for `LIVE` and `https://sandbox.gateway.sezzle.com/v1` for using `Sezzle` in `Sandbox` mode.
* Save the settings.

### Your store is now ready to use Sezzle as a payment gateway.

## Restrict Sezzle based on user roles
Make sure `Sezzle` plugin is `active` in `Wordpress` admin.

#### Hide Sezzle Payment Gateway
If you want to hide` Sezzle's` payment gateway based on user roles

1. Add the following function to your code:

```php
function restrict_sezzle_pay($available_gateways) {
    unset($available_gateways['sezzlepay']);
    return $available_gateways;
}
```

2. Call the following filter `inside` the user's access deciding code:

```php
add_filter('woocommerce_available_payment_gateways', 'restrict_sezzle_pay');
```

## Frontend Functonality

* If you have correctly set up `Sezzle`, you will see `Sezzle` as a payment method in the checkout page.
* Select `Sezzle` and move forward.
* Once you click `Place Order`, you will be redirected to `Sezzle Checkout` to complete the checkout and eventually in `WooCommerce` too.

## Capture Payment

* Capture is performed instantly from the plugin after order is created and validated in WooCommerce.

## Refund Payment

* Go to `WooCommerce > Orders` after logging into Wordpress admin.
* Go inside the order that you want to refund and click `Refund`.
* Provide the amount you want to refund in the respective places and click on `Refund $xx.xx via Sezzle`.
* If refund is successful, you will see `Order Status` as either `Refunded` or `Processing`(in case of `Partial Refund`) as per the input given and an `Order Note` added as `Refund of xx.xx successfully sent to Sezzle`.
* The same can be verified in the `Sezzle Merchant Dashboard`. `Status` as `Refunded` indicates the payment has been fully refunded and `Status` as `Partially Refunded` indicates the payment has been partially refunded.  

## Order Verification in WooCommerce Admin

* Login to `Wordpress` admin and navigate to `WooCommerce > Orders`.
* Proceed into the corresponding order.
* `Order Status` as `Proceesing` and `Order Note` shown as `Payment approved by Sezzle` successfully means payment is successfully captured by `Sezzle`.
* `Order Status` as `Pending` and `Order Note` not shown as `Payment approved by Sezzle` successfully means there is something wrong while processing the order.

## Order Verification in Sezzle Merchant Dashboard

* Login to `Sezzle Merchant Dashboard` and navigate to `Orders`.
* Proceed into the corresponding order.
* Status as `Approved` means payment is successfully captured by `Sezzle`.
* Status as `Authorized`, uncaptured means payment is authorized but yet not captured.

## How Sandbox works?

* In the `Sezzle` configuration page of your `WooCommerce` admin, enter the `Sandbox` `API Keys` from your [`Sezzle Merchant Sandbox Dashboard`](https://sandbox.dashboard.sezzle.com/merchant/) and set the `Sezzle API URL` to `https://sandbox.gateway.sezzle.com/v1`, then save the configuration. Make sure you are doing this on your `dev/staging` website.
* On your website, add an item to the cart, then proceed to `Checkout` and select `Sezzle` as the payment method.
* Click `Continue` then `Place Order` and you should be redirected to the `Sezzle Checkout` page. If prompted, sign in and continue.
* Enter the payment details using test data, then click `Complete Order`.
* After the payment is completed on `Sezzle`, you should be redirected back to your website and see a successful payment page.
* `Sandbox` testing is complete. You can login to your `Sezzle Merchant Sandbox Dashboard` to see the test order you just placed.

## Troubleshooting/Debugging

* There is logging enabled by `Sezzle` for tracing the `Sezzle` actions.
* In case merchant is facing issues which is unknown to `Merchant Success` and `Support` team, they can ask for this logs and forward to the `Platform Integrations` team.
* Name of the log should be like `Sezzlepay-2020-04-27-ec6929b2d9c82df755ba0835a5ce6337.log`.


changelog.txt
*** Sezzle Payment Changelog ***

2021-02-26 - Version 4.0.3
* Fix: Readme file changes 

2021-01-01 - Version 4.0.2
* Feature: Multisite Support

2020-12-04 - Version 3.1.4
* FIX: Sqaure and Stripe Payment Method Form blocking.
* FEATURE: Ability to turn on/off installment widget plan from Sezzle settings.

2020-11-26 - Version 3.1.3
* FIX: Multiple Installment Widget.

2020-11-25 - Version 3.1.2
* FEATURE: Installment Plan Widget under Sezzle Payment Option in Checkout Page.
* FIX: Admin check added in gateway hiding function.

2020-08-31 - Version 3.1.1
* FIX: Failing of sudden orders being already captured.
* FEATURE: Ability to turn on/off logging.

2020-08-31 - Version 3.1.0
* MODIFY: Transaction Mode added instead of Sezzle API URL. Now, user can select between LIVE and SANDBOX mode instead of adding the URL.

2020-07-02 - Version 3.0.5
* FIX: Undefined index:Authorization during redirection to Sezzle.

2020-05-12 - Version 3.0.4
* MODIFY: Updated User Guide.

2020-02-19 - Version 3.0.3
* MODIFY: Updated Widget Script URL.

2019-12-20 - Version 3.0.2
* FIX: Order property access through function instead of direct access.

2019-11-19 - Version 3.0.1
* FIX: Return URL from Sezzle Checkout changed to Checkout URL of merchant website.
* FEATURE: Added logs for checking API functions.
* FIX: Check payment capture status before capturing the payment so that already captured orders does not fall into the process.

2019-09-13 - Version 3.0.0
* FIX: Downgraded to previous stable version due to some conflicts arising in few versions.
* MODIFY: Delayed capture has been removed.
* MODIFY: Widget in Cart has been removed.

2019-09-03 - Version 2.0.9
* FIX: Added check to include settings class when not available.

2019-08-28 - Version 2.0.8
* MODIFY: Wordpress support version has been changed to 4.4.0 or higher.

2019-08-28 - Version 2.0.7
* FEATURE: Hiding of Sezzle Pay based on cart total.
* FEATURE: Sezzle Widget and Sezzle Payment merged into one plugin.
* FIX : Amount converted to cents while refund.

2019-08-21 - Version 2.0.6
* FIX: Page hanging issue during order status change for other payment methods.

2019-08-14 - Version 2.0.5
* FIX: Security fix and quality improvements.

2019-08-12 - Version 2.0.4
* FEATURE: Delayed Capture.
* FEATURE: Sezzle Widget for Cart Page.
* FEATURE: New settings for managing Sezzle Widget.


internal/scripts/install.sh
#!/bin/sh

mysql_ready() {
    mysqladmin ping -h $MYSQL_HOST -u $MYSQL_USER -p$MYSQL_PASSWORD > /dev/null 2>&1
}

while !(mysql_ready)
do
    sleep 3
    echo "waiting for mysql database to be ready..."
done

# wp core download --version=$WORDPRESS_VERSION --locale=en_US
wp config create \
    --dbhost=$MYSQL_HOST \
    --dbname=$MYSQL_DATABASE \
    --dbuser=$MYSQL_USER \
    --dbpass="$MYSQL_PASSWORD" \
    --skip-check \
    --allow-root \
    --extra-php <<PHP 
    $WORDPRESS_CONFIG_EXTRA 
PHP

wp core $([ "$WORDPRESS_MULTISITE" = true ] && echo "multisite-install" || echo "install") \
    --url=http://localhost \
    --title="LUMA" \
    --admin_user=$WORDPRESS_ADMIN_USERNAME \
    --admin_password=$WORDPRESS_ADMIN_PASSWORD \
    --admin_email=$WORDPRESS_ADMIN_EMAIL \
    --skip-email \
    --allow-root

wp plugin install wordpress-importer --activate
wp plugin install woocommerce --version=$WOOCOMMERCE_VERSION --allow-root --activate
wp theme install storefront --allow-root --activate
php -f woocommerce-setup-wizard.php
wp import ./woocommerce-products.xml --authors=create --quiet --allow-root

wp wc customer create \
    --user=1 \
    --email="$WOOCOMMERCE_CUSTOMER_EMAIL" \
    --password="$WOOCOMMERCE_CUSTOMER_PASSWORD" \
    --username="$WOOCOMMERCE_CUSTOMER_USERNAME" \
    --first_name='Sezzle' \
    --last_name='Customer' \
    --billing='{"first_name":"Sezzle","last_name":"Customer","phone":"9000090000","address_1":"91Springboard","address_2":"JP Nagar 4th Phase","city":"Bengaluru","state:":"KA","country":"IN","postcode":"560076"}' \
    --shipping='{"first_name":"Sezzle","last_name":"Customer","phone":"9000090000","address_1":"91Springboard","address_2":"JP Nagar 4th Phase","city":"Bengaluru","state:":"KA","country":"IN","postcode":"560076"}'

rm -rf wp-config-sample.php wp-admin/install*.php

internal/scripts/woocommerce-setup-wizard.php
<?php

echo "Auto setup WooCommerce :)\n\n";
require_once("./wp-load.php");

echo "Setup default values...\n";
$address        	= "Arekere";
$address_2      	= "Arekere";
$city           	= "Bengaluru";
$country_state  	= "KA:IN";
$postcode       	= "560076";
$currency_code  	= "INR";
$product_type   	= "both";
$tracking       	= false;
$currency_pos 		= 'left';
$price_decimal_sep  = '.';
$price_num_decimals = '2';
$price_thousand_sep = ',';

echo "Updating options...\n";
update_option('woocommerce_store_address', $address);
update_option('woocommerce_store_address_2', $address_2);
update_option('woocommerce_store_city', $city);
update_option('woocommerce_default_country', $country_state);
update_option('woocommerce_store_postcode', $postcode);
update_option('woocommerce_currency', $currency_code);
update_option('woocommerce_product_type', $product_type);
update_option('woocommerce_currency_pos', $currency_pos);
update_option('woocommerce_price_decimal_sep', $price_decimal_sep);
update_option('woocommerce_price_num_decimals', $price_num_decimals);
update_option('woocommerce_price_thousand_sep', $price_thousand_sep);
update_option('woocommerce_store_address', $address);
update_option('woocommerce_admin_notices', 'a:0:{}');
update_option('woocommerce_flat_rate_1_settings','a:3:{s:5:"title";s:9:"Flat rate";s:10:"tax_status";s:7:"taxable";s:4:"cost";s:1:"5";}');
update_option('woocommerce_flat_rate_2_settings1','a:3:{s:5:"title";s:9:"Flat rate";s:10:"tax_status";s:7:"taxable";s:4:"cost";s:2:"30";}');
update_option('permalink_structure', '/%year%/%monthnum%/%day%/%postname%/');
update_option('woocommerce_permalinks','a:4:{s:13:"category_base";s:0:"";s:8:"tag_base";s:0:"";s:14:"attribute_base";s:0:"";s:12:"product_base";s:0:"";}');

echo "Updating tracking...\n";
if ($tracking) 
{
	update_option('woocommerce_allow_tracking','yes');
	WC_Tracker::send_tracking_data(true);
}
else 
{
	update_option('woocommerce_allow_tracking','no');
}

echo "Create pages...\n";
WC_Install::create_pages();

echo "Updating home page...\n";
$page = get_page_by_title('Shop');
update_option('show_on_front','page');
update_option('page_on_front', $page->ID);

echo "DONE! :)";
?>